proposals/0000-pure-template-haskell.md

```diff
@@ -0,0 +1,294 @@
+
+## Proposed Change Specification
+
+### 0. Terminology
+
+A Template Haskell quote is said to be "impure" if it has type `m a` where `m` has or requires a `MonadIO` instance.
+
+A Template Haskell splice is said to be "impure" if the code inside the splice has type `m a` where `m` has or requires a `MonadIO` instance.
+
+Note that today types for quotes and splice code are already inferred with appropriate constraints depending on what they use.
+
+### 1. Syntax
+
+A new syntax for top-level “impure splices” is introduced.
+
+- Impure untyped splice: `$!..`
+- Impure typed splice: `$$!..`
+
+We will elide the distinction between a splice being impure in the sense of "needing IO" and a splice being impure in the sense of "using impure splice syntax", so long as this is not confusing.
+
+Splices which do not use the impure splice syntax are called “normal” splices. This includes:
+- Splices of all kinds when `EnforceTemplateHaskellPurity` is disabled
+- Pure splices when `EnforceTemplateHaskellPurity` is enabled
+- Nested (non-top-level) splices
+
+### 2. Additions to `template-haskell`
+
+Introduce a new class `Reify`, containing the `reify` methods from `Quasi`:
+```
+–- could use MonadError here, but we’re going to impose a Reify m constraint at 
+-- the top level, and we probably want users to be able to pick their own 
+-- error monad for m if necessary
+class Monad m => Reify m where
+  reify          :: Name -> ExceptT ReifyError m Info
+  reifyFixity    :: Name -> ExceptT ReifyError m (Maybe Fixity)
+  reifyType      :: Name -> ExceptT ReifyError m Type
+  reifyInstances :: Name -> [Type] -> ExceptT ReifyError m [Dec]
+  reifyRoles         :: Name -> ExceptT ReifyError m [Role]
+  reifyAnnotations   :: Data a => AnnLookup -> ExceptT ReifyError m [a]
+  reifyModule        :: Module -> ExceptT ReifyError m ModuleInfo
+  reifyConStrictness :: Name -> ExceptT ReifyError m [DecidedStrictness]
+```
+
+Introduce a type `PureQ`, which parallels `Q` but only has access to `Quote` and `Reify`.
+```
+newtype PureQ a = PureQ { unPureQ :: forall m . (Quote m, Reify m) => m a }
+```
+
+`PureQ` does not provide an explicit error-reporting mechanism. For reporting errors to the user, splices should still throw exceptions, which can be caught by the compiler and reported to the user.
+
+### 3. Language extension
+
+Impure splice syntax is permitted only if the language extension `ImpureTemplateHaskell` is enabled.
+The code in an impure splice must have type `Q a`, as with normal splices today. 
+
+The behaviour of normal splices is controlled with a new language extension, `EnforceTemplateHaskellPurity`. 
+The effect of `EnforceTemplateHaskellPurity` is that the code in a top-level normal splice must have type `PureQ a` instead of `Q a`.
+
+Additionally, there is a new language extension `PureTemplateHaskell`, which implies `NoImpureTemplateHaskell` and `EnforceTemplateHaskellPurity`.
```

Comment:
Good questions.
- `TemplateHaskellQuotes` is I think independent, since this is only really about splices.
- I don't think `PureTemplateHaskell` should imply `TemplateHaskell`. The reason being that you might want to set `PureTemplateHaskell` in your `default-extensions`, or it might become part of a language edition. But I think you don't want to enable `TemplateHaskell` everywhere, since I _think_ (maybe I'm wrong?) that the mere presence of that extension causes GHC to treat the module as if it has TH in it, which would be bad. I think it's useful to be able to express "_if_ you use TH, it's pure" as a thing separate from "I am using TH".